ar71xx: gl-ar300m nand support

hanetzer opened the pull request:
Currently working on porting/rebasing/sanitizing/etc the vendor support for the
nand side of the gl-ar300m, based on domino-team/lede-ar300m@333ccb0e158edaf80cb1ca696e328f9435f7d3eb and
onward. I've rebased it and cleaned it up as much as possible without messing
with its code too much yet, but the `ath79_spinand.c` driver will require some
work in porting it from 4.4 to 4.9's kernel mtd/nand api.

Mostly opening this as a place of discussion, visibility and q&a.

diizzyy commented on target/linux/ar71xx/image/nand.mk:
```diff
@@ -57,6 +57,18 @@ define Device/hiveap-121
 
+#define Device/GL-AR300MD
+  #DEVICE_TITLE := GL AR300M(Dual)
+  #DEVICE_PACKAGES := kmod-usb-core kmod-usb2 kmod-usb-storage kmod-usb-ledtrig-usbport kmod-ath10k ath10k-firmware-qca988x
```
According to GL.iNet it's only single band so you should only need ath9k, GL-AR300MD is dual band however.
https://www.gl-inet.com/product/gl-ar300m/

pepe2k commented:
Hi @hanetzer,

Please, have a look at #272.

Cheers,
Piotr

hanetzer commented:
@pepe2k aware thiw will require changes, thanks for pointing me at that older pr, will be a good reference.

hanetzer commented:
@pepe2k with regards to the staging spi nand driver, should I use it at the `drivers/staging` location in tree or go ahead and relocate it to say `files/drivers/mtd/nand/` or the like? I've just picked it back up atm and got some tinkerin' to do.

hanetzer commented:
Progress has been made; I'm using the mt29f spinand driver, but it seems the nand flash I have is an odd one (different hardware versions and such), gotta iron out the kinks in that, and hopefully find a real manufacturer name and datasheet for it to get it properly supported.

hanetzer commented:
Current status:
```
Loading from device 0: ath-spi-nand (offset 0x0)
   Image Name:   MIPS OpenWrt Linux-4.9.70
   Created:      2017-12-24   8:21:08 UTC
   Image Type:   MIPS Linux Kernel Image (lzma compressed)
   Data Size:    1550246 Bytes =  1.5 MB
   Load Address: 80060000
   Entry Point:  80060000
## Booting image at 81000000 ...
   Image Name:   MIPS OpenWrt Linux-4.9.70
   Created:      2017-12-24   8:21:08 UTC
   Image Type:   MIPS Linux Kernel Image (lzma compressed)
   Data Size:    1550246 Bytes =  1.5 MB
   Load Address: 80060000
   Entry Point:  80060000
   Verifying Checksum at 0x81000040 ...OK
   Uncompressing Kernel Image ... OK
No initrd
## Transferring control to Linux (at address 80060000) ...
## Giving linux memsize in bytes, 134217728

Starting kernel ...

[    0.000000] Linux version 4.9.70 (hanetzer@proprietary-killer) (gcc version 5.5.0 (OpenWrt GCC 5.5.0 r5489-ec78d6cbe5) ) #0 Sun Dec 24 08:21:08 2017
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019374 (MIPS 24Kc)
[    0.000000] SoC: Qualcomm Atheros QCA9533 ver 2 rev 0
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 08000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] OF: fdt:No valid device tree found, continuing without
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Built 1 zonelists in Zone order, mobility grouping on.  Total pages: 32512
[    0.000000] Kernel command line:  board=GL-AR300M mtdparts=spi0.0:256k(u-boot)ro,64k(u-boot-env),16000k(reserved),64k(art)ro;spi0.1:2048k(kernel),-(ubi) console=ttyS0,115200 rootfstype=squashfs noinitrd
[    0.000000] PID hash table entries: 512 (order: -1, 2048 bytes)
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Writing ErrCtl register=00000000
[    0.000000] Readback ErrCtl register=00000000
[    0.000000] Memory: 124360K/131072K available (3597K kernel code, 183K rwdata, 908K rodata, 232K init, 212K bss, 6712K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS:51
[    0.000000] Clocks: CPU:650.000MHz, DDR:597.778MHz, AHB:216.666MHz, Ref:25.000MHz
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 5880801374 ns
[    0.000009] sched_clock: 32 bits at 325MHz, resolution 3ns, wraps every 6607641598ns
[    0.008234] Calibrating delay loop... 432.53 BogoMIPS (lpj=2162688)
[    0.071098] pid_max: default: 32768 minimum: 301
[    0.076072] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.083050] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.094629] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.105033] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.112008] NET: Registered protocol family 16
[    0.118069] MIPS: machine is GL-AR300M
[    0.603153] ar724x-pci ar724x-pci.0: PCIe link is down
[    0.608614] registering PCI controller with io_map_base unset
[    0.626865] PCI host bridge to bus 0000:00
[    0.631249] pci_bus 0000:00: root bus resource [mem 0x10000000-0x11ffffff]
[    0.638492] pci_bus 0000:00: root bus resource [io  0x0000]
[    0.644368] pci_bus 0000:00: root bus resource [??? 0x00000000 flags 0x0]
[    0.651523] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]
[    0.660562] clocksource: Switched to clocksource MIPS
[    0.666906] NET: Registered protocol family 2
[    0.672323] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    0.679679] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.686441] TCP: Hash tables configured (established 1024 bind 1024)
[    0.693260] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.699422] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.706299] NET: Registered protocol family 1
[    0.714628] Crashlog allocated RAM at address 0x3f00000
[    0.721067] workingset: timestamp_bits=30 max_order=15 bucket_order=0
[    0.734447] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.740635] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.758563] io scheduler noop registered
[    0.762747] io scheduler deadline registered (default)
[    0.768630] Serial: 8250/16550 driver, 1 ports, IRQ sharing disabled
[    0.775766] console [ttyS0] disabled
[    0.799665] serial8250.0: ttyS0 at MMIO 0x18020000 (irq = 11, base_baud = 1562500) is a 16550A
[    0.808771] console [ttyS0] enabled
[    0.808771] console [ttyS0] enabled
[    0.816334] bootconsole [early0] disabled
[    0.816334] bootconsole [early0] disabled
[    0.828245] m25p80 spi0.0: found w25q128, expected m25p80
[    0.833891] m25p80 spi0.0: w25q128 (16384 Kbytes)
[    0.838784] 4 cmdlinepart partitions found on MTD device spi0.0
[    0.844911] Creating 4 MTD partitions on "spi0.0":
[    0.849867] 0x000000000000-0x000000040000 : "u-boot"
[    0.856777] 0x000000040000-0x000000050000 : "u-boot-env"
[    0.863826] 0x000000050000-0x000000ff0000 : "reserved"
[    0.870628] 0x000000ff0000-0x000001000000 : "art"
[    0.877693] libphy: Fixed MDIO Bus: probed
[    0.896630] libphy: ag71xx_mdio: probed
[    1.533153] ag71xx ag71xx.0: connected to PHY at ag71xx-mdio.1:04 [uid=004dd042, driver=Generic PHY]
[    1.543294] eth0: Atheros AG71xx at 0xb9000000, irq 4, mode:MII
[    2.173105] ag71xx-mdio.1: Found an AR934X built-in switch
[    2.226439] eth1: Atheros AG71xx at 0xba000000, irq 5, mode:GMII
[    2.241664] nand: device found, Manufacturer ID: 0xa1, Chip ID: 0xe1
[    2.248237] nand: FW Device PN256GO1 1G 3.3V 8-bit
[    2.253206] nand: 128 MiB, SLC, erase size: 128 KiB, page size: 2048, OOB size: 128
[    2.261156] nand: WARNING: spi0.1: the ECC used on your system is too weak compared to the one required by the NAND chip
[    2.272392] Scanning device for bad blocks
[    2.545306] Bad eraseblock 1022 at 0x000007fc0000
[    2.550360] Bad eraseblock 1023 at 0x000007fe0000
[    2.555230] random: fast init done
[    2.558754] 2 cmdlinepart partitions found on MTD device spi0.1
[    2.564879] Creating 2 MTD partitions on "spi0.1":
[    2.569836] 0x000000000000-0x000000200000 : "kernel"
[    2.576373] 0x000000200000-0x000008000000 : "ubi"
[    2.584236] NET: Registered protocol family 10
[    2.591285] NET: Registered protocol family 17
[    2.596028] 8021q: 802.1Q VLAN Support v1.8
[    2.605636] UBI: auto-attach mtd5
[    2.609092] ubi0: attaching mtd5
[    2.612672] ubi0 error: validate_ec_hdr: bad VID header offset 2048, expected 512
[    2.620402] ubi0 error: validate_ec_hdr: bad EC header
[    2.625731] Erase counter header dump:
[    2.629599]  magic          0x55424923
[    2.633482]  version        1
[    2.636542]  ec             0
[    2.639606]  vid_hdr_offset 2048
[    2.642944]  data_offset    4096
[    2.646274]  image_seq      1548887763
[    2.650143]  hdr_crc        0x1229687f
[    2.654027] erase counter header hexdump:
[    2.658192] CPU: 0 PID: 1 Comm: swapper Not tainted 4.9.70 #0
[    2.664136] Stack : 805377fa 00000031 00000000 00000001 87c2c284 804cf247 8046b070 00000001
[    2.672846]         80533940 1229687f 00000000 87dc9b14 87ddc100 800aa3f4 30203030 20303020
[    2.681547]         30302030 30203030 8046f150 87c31c64 39203638 800da0dc 202e2e2e 2e2e2e2e
[    2.690237]         2e2e2e2e 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    2.698947]         00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    2.707647]         ...
[    2.710184] Call Trace:
[    2.712732] [<8006b288>] show_stack+0x54/0x88
[    2.717238] [<802a7b4c>] validate_ec_hdr+0xdc/0xfc
[    2.722205] [<802a880c>] ubi_io_read_ec_hdr+0x270/0x2c0
[    2.727601] [<802ae18c>] ubi_attach+0x23c/0x1480
[    2.732383] [<802a1d78>] ubi_attach_mtd_dev+0x8f4/0xdc4
[    2.737787] [<8050efb8>] ubi_init+0x284/0x340
[    2.742300] [<80060a70>] do_one_initcall+0xdc/0x1ac
[    2.747338] [<804f6d50>] kernel_init_freeable+0x168/0x220
[    2.752931] [<803de114>] kernel_init+0x10/0x104
[    2.757617] [<800660d8>] ret_from_kernel_thread+0x14/0x1c
[    2.763209] ubi0 error: ubi_io_read_ec_hdr: validation failed for PEB 0
[    2.770049] ubi0 error: ubi_attach_mtd_dev: failed to attach mtd5, error -22
[    2.777372] UBI error: cannot attach mtd5[    2.781727] VFS: Cannot open root device "(null)" or unknown-block(0,0): error -6
[    2.789456] Please append a correct "root=" boot option; here are the available partitions:
[    2.798132] 1f00             256 mtdblock0 [    2.802281]  (driver?)
[    2.804719] 1f01              64 mtdblock1 [    2.808852]  (driver?)
[    2.811310] 1f02           16000 mtdblock2 [    2.815446]  (driver?)
[    2.817889] 1f03              64 mtdblock3 [    2.822038]  (driver?)
[    2.824475] 1f04            2048 mtdblock4 [    2.828609]  (driver?)
[    2.831067] 1f05          129024 mtdblock5 [    2.835204]  (driver?)
[    2.837645] Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0)
[    2.847362] Rebooting in 1 seconds..```

mkresin commented:
Looks like the pagesize passed to the append-ubi build step is wrong (2048 vs. 512).

```
git grep MKUBIFS_OPTS
git grep UBIFS_OPTS
```

for more details.

hanetzer commented:
@mkresin ok, I've made some adjustments to the device definition in nand.mk, this flashes and boots. If there is a better way to do what I have done, let me know.

pepe2k commented:
Superseded by https://github.com/openwrt/openwrt/pull/615.
